tables/TaQL/ExprNodeSetOpt.cc

```diff
@@ -0,0 +1,450 @@
+            rightC = set[inx]->isRightClosed();
+          }
+        } else {
+          // No overlap, so create the interval found and start a new one.
+          newStart.push_back (stval);
+          newEnd.push_back (endval);
+          newLeftC.push_back (leftC);
+          newRightC.push_back (rightC);
+          rowNrs.push_back (rownr);
+          stval  = st2;
+          endval = end2;
+          leftC  = set[inx]->isLeftClosed();
+          rightC = set[inx]->isRightClosed();
+          rownr  = inx;
+        }
+      }
+      // Create the last interval.
+      newStart.push_back (stval);
+      newEnd.push_back (endval);
+      newLeftC.push_back (leftC);
+      newRightC.push_back (rightC);
+      rowNrs.push_back (rownr);
+    }
+    // Now create the correct object.
+    return createOptSet (set, newStart, newEnd, newLeftC, newRightC);
+  }
+
+  template<typename T>
+  TENShPtr TableExprNodeSetOptContSet<T>::createOptSet
+  (const TableExprNodeSet& set,
+   const std::vector<T>& start, const std::vector<T>& end, 
+   const std::vector<Bool>& leftC, const std::vector<Bool>& rightC)
+   {
+    // See if all intervals have the same left and right closedness.
+    // If so, a better version can be used that does not need to test on it
+    // for each compare.
+    Bool same = True;
+    for (size_t i=1; i<leftC.size(); ++i) {
+      if (leftC[i] != leftC[0]  ||  rightC[i] != rightC[0]) {
+        same = False;
+        break;
+      }
+    }
+    // Create the appropriate object.
+    TableExprNodeSetOptContSetBase* optSet;
+    if (same) {
+      if (leftC[0]) {
+        if (rightC[0]) {
+          optSet = new TableExprNodeSetOptContSetCC<T> (set, start, end);
+        } else {
+          optSet = new TableExprNodeSetOptContSetCO<T> (set, start, end);
+        }
+      } else {
+        if (rightC[0]) {
+          optSet = new TableExprNodeSetOptContSetOC<T> (set, start, end);
+        } else {
+          optSet = new TableExprNodeSetOptContSetOO<T> (set, start, end);
+        }
+      }
+    } else {
+      optSet = new TableExprNodeSetOptContSet<T> (set, start, end,
+                                                  leftC, rightC);
+    }
+    TENShPtr ptr(optSet);
+    ///optSet->setRows (rowNrs);
+    return ptr;
+  }
+
+
+  template<typename T>
+  TableExprNodeSetOptContSetCC<T>::TableExprNodeSetOptContSetCC
+  (const TableExprNodeSet& orig,
+   const std::vector<T>& starts,
+   const std::vector<T>& ends)
+    : TableExprNodeSetOptContSet<T> (orig, starts, ends)
+  {}
+
+  template<typename T>
+  void TableExprNodeSetOptContSetCC<T>::show (ostream& os, uInt indent) const
+  {
+    TableExprNodeSetOptContSet<T>::show (os, indent);
+    os << " as TableExprNodeSetOptContSetCC" << endl;
+  }
+  
+  template<typename T>
+  Int64 TableExprNodeSetOptContSetCC<T>::find (T value) const
+  {
+    // Use lambda function to use <= instead of <.
+    auto iter = std::upper_bound (this->itsEnds.begin(), this->itsEnds.end(), value,
```

Comment:
This does not explain why `<=` is the proper operation.
It also doesn't explain why `std::less_equal` can't be used.

Reply:
Now using std::less_equal. The interval is closed on the right side, thus needs <=.